File: src/main/java/com/paragon/client/systems/module/hud/HUDModule.java
package com.paragon.client.systems.module.hud;

import com.paragon.api.util.render.GuiUtil;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.api.util.render.TextRenderer;
import com.paragon.client.systems.module.Module;
import com.paragon.client.systems.module.Category;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.util.math.MathHelper;

import java.awt.*;

public abstract class HUDModule extends Module implements TextRenderer {

    private float x = 50, y = 50;
    private float lastX, lastY;
    private boolean dragging;

    public HUDModule(String name, String description) {
        super(name, Category.HUD, description);
        this.setVisible(false);
    }

    public abstract void render();
    public abstract float getWidth();
    public abstract float getHeight();

    public void updateComponent(int mouseX, int mouseY) {
        // Set X and Y
        if (dragging) {
            ScaledResolution sr = new ScaledResolution(mc);

            float newX = MathHelper.clamp(mouseX - lastX, 4, RenderUtil.getScreenWidth() - getWidth());
            float newY = MathHelper.clamp(mouseY - lastY, 4, RenderUtil.getScreenHeight() - getHeight());

            this.x = newX;
            this.y = newY;

            float centerX = newX + (getWidth() / 2f);
            float centerY = newY + (getHeight() / 2f);

            if (centerX > (sr.getScaledWidth() / 2f) - 5 && centerX < (sr.getScaledWidth() / 2f) + 5) {
                this.x = (sr.getScaledWidth() / 2f) - (getWidth() / 2f);
            }

            if (centerY > (sr.getScaledHeight() / 2f) - 5 && centerY < (sr.getScaledHeight() / 2f) + 5) {
                this.y = (sr.getScaledHeight() / 2f) - (getHeight() / 2f);
            }
        }
    }

    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
        if (GuiUtil.mouseOver(getX(), getY(), getX() + getWidth(), getY() + getHeight(), mouseX, mouseY)) {
            if (mouseButton == 0) {
                this.lastX = mouseX - getX();
                this.lastY = mouseY - getY();

                dragging = true;
            } else if (mouseButton == 1) {
                if (this.isEnabled()) {
                    toggle();
                }
            }
        }
    }

    public void mouseReleased(int mouseX, int mouseY, int mouseButton) {
        dragging = false;
    }

    public float getX() {
        return x;
    }

    public void setX(float newX) {
        this.x = newX;
    }

    public float getY() {
        return y;
    }

    public void setY(float newY) {
        this.y = newY;
    }

    public boolean isDragging() {
        return dragging;
    }

}

File: src/main/kotlin/com/paragon/impl/module/hud/HUDModule.kt
package com.paragon.impl.module.hud

import com.paragon.impl.module.Category
import com.paragon.impl.module.Module
import com.paragon.impl.module.annotation.NotVisibleByDefault
import com.paragon.impl.module.hud.impl.HUDEditor
import com.paragon.impl.setting.Setting
import com.paragon.util.render.font.FontUtil
import com.paragon.util.roundToNearest
import net.minecraft.client.gui.ScaledResolution
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

/**
 * @author Surge, SooStrator1136
 */
@SideOnly(Side.CLIENT)
@NotVisibleByDefault
abstract class HUDModule(name: String, description: String) : Module(name, Category.HUD, description) {

    val alignment = Setting("Alignment", FontUtil.Align.LEFT) describedBy "Align text HUDs"

    open var width = 50F
    open var height = 50F

    var x = 50F
    var y = 50F
    private var lastX = 0F
    private var lastY = 0F
    var isDragging = false
        private set

    abstract fun render()

    fun updateComponent(mouseX: Int, mouseY: Int) {
        // Set X and Y
        if (isDragging) {
            val sr = ScaledResolution(minecraft)
            val newX = mouseX - lastX
            val newY = mouseY - lastY

            x = newX
            y = newY

            val centerX = newX + width / 2f
            val centerY = newY + height / 2f

            if (centerX > sr.scaledWidth / 2f - 5 && centerX < sr.scaledWidth / 2f + 5) {
                x = sr.scaledWidth / 2f - width / 2f
            }

            if (centerY > sr.scaledHeight / 2f - 5 && centerY < sr.scaledHeight / 2f + 5) {
                y = sr.scaledHeight / 2f - height / 2f
            }

            x = x.roundToNearest(HUDEditor.snap.value.toFloat()).toFloat()
            y = y.roundToNearest(HUDEditor.snap.value.toFloat()).toFloat()
        }
    }

    fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int): Boolean {
        if (isHovered(x - when (alignment.value) {
                FontUtil.Align.CENTER -> width / 2f
                FontUtil.Align.RIGHT -> width
                else -> 0f
            }, y, width, height, mouseX, mouseY)) {
            if (mouseButton == 0) {
                lastX = mouseX - x
                lastY = mouseY - y
                isDragging = true
            }
            else if (mouseButton == 1) {
                if (isEnabled) {
                    toggle()
                    return true
                }
            }
            else if (mouseButton == 2) {
                alignment.setValue(alignment.nextMode)
                return true
            }
        }
        return false
    }

    fun mouseReleased(mouseX: Int, mouseY: Int, mouseButton: Int) {
        isDragging = false
    }

}